service/authz.go

```diff
@@ -0,0 +1,113 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//           $$\                           $$\ $$\   $$\
+//           $$ |                          $$ |\__|  $$ |
+//  $$$$$$$\ $$ | $$$$$$\  $$\   $$\  $$$$$$$ |$$\ $$$$$$\    $$$$$$\   $$$$$$\
+// $$  _____|$$ |$$  __$$\ $$ |  $$ |$$  __$$ |$$ |\_$$  _|  $$  __$$\ $$  __$$\
+// $$ /      $$ |$$ /  $$ |$$ |  $$ |$$ /  $$ |$$ |  $$ |    $$ /  $$ |$$ | \__|
+// $$ |      $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$\ $$ |  $$ |$$ |
+// \$$$$$$\  $$ |\$$$$$   |\$$$$$   |\$$$$$$  |$$ |  \$$$   |\$$$$$   |$$ |
+//  \_______|\__| \______/  \______/  \_______|\__|   \____/  \______/ \__|
+//
+// This file is part of Clouditor Community Edition.
+
+package service
+
+import (
+	"context"
+
+	"clouditor.io/clouditor/api/orchestrator"
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/golang-jwt/jwt/v4"
+	grpc_auth "github.com/grpc-ecosystem/go-grpc-middleware/auth"
+)
+
+type RequestType int
+
+const (
+	AccessCreate RequestType = iota
+	AccessRead
+	AccessUpdate
+	AccessDelete
+)
+
+// AuthorizationStrategy is an interface that implements a function which whether the current cloud service request can
+// be fulfilled using the current access strategy.
+type AuthorizationStrategy interface {
+	CheckAccess(ctx context.Context, typ RequestType, req orchestrator.CloudServiceRequest) bool
+	AllowedCloudServices(ctx context.Context) (all bool, IDs []string)
+}
+
+// AuthorizationStrategyJWT is an AuthorizationStrategy that expects a list of cloud service IDs to be in a specific JWT
+// claim key.
+type AuthorizationStrategyJWT struct {
+	Key string
+}
+
+// CheckAccess checks whether the current request can be fulfilled using the current access strategy.
+func (a *AuthorizationStrategyJWT) CheckAccess(ctx context.Context, typ RequestType, req orchestrator.CloudServiceRequest) bool {
+	var list []string
+
+	// Retrieve the list of allowed cloud services. we never allow to retrieve
+	// "all" services with the token strategy.
+	_, list = a.AllowedCloudServices(ctx)
+
+	return slices.Contains(list, req.GetCloudServiceId())
+}
+
+// AllowedCloudServices retrieves a list of allowed cloud service IDs according to the current access strategy.
+func (a *AuthorizationStrategyJWT) AllowedCloudServices(ctx context.Context) (all bool, list []string) {
+	var err error
+	var token string
+	var claims jwt.MapClaims
+
+	// Retrieve the raw token from the context
+	token, err = grpc_auth.AuthFromMD(ctx, "bearer")
+	if err != nil {
+		return false, nil
+	}
+
+	// We need to re-parse the already validated claim to get a specific key from the claims map.
+	parser := jwt.NewParser()
+	_, _, err = parser.ParseUnverified(token, &claims)
+	if err != nil {
+		return false, nil
+	}
+
+	for _, item := range claims[a.Key].([]interface{}) {
+		list = append(list, item.(string))
+	}
+
+	return false, list
```

Comment:
Your suggestion won't work. The contents of `claims["myKey"]` is a `[]interface{}` which then contains `string` elements. Theoretically it could even contain string and non-string elements. So I need to loop through it, but casting to `[]string` won't work. I will add some type assertions.

Reply:
But then line 120 would panic, if current item in non-string

Reply:
I added an additional check in the for-loop.